### PR TITLE
focal: add python3.8 slices

### PR DESCRIPTION
**Pull request by rebornplusplus**
This PR adds the necessary slices for a chiselled python3.8.

The slices should support almost all of the python3.8 standard library, except for the following modules:
- `ensurepip` (will need the `python3-pip` package)
- `tkinter`, `tkinter.*` (will need the `python3-tk` package)
   * `turtle` (although this module is included, it directly depends on `tkinter`)
- `msilib`, `msvcrt`, `winreg`, `winsound` (These are MS Windows specific services and only available on MS Windows platforms)

Additionally, I intentionally did not create any `/usr/bin/python3` symlink for any possible conflicts with other future versions. Let me know if you think I should have that pointing to `/usr/bin/python3.8` anyway. Cheers!

---

<details>
  <summary>python3.8 package dependencies</summary>
  
  Here goes a trimmed-down, relevant dependency of the `python3.8` package on Focal. The full list can be found [here as a GitHub Gist](https://gist.github.com/rebornplusplus/683b4010870a70db918051f1e14afc4b).
  
```
python3.8
  Depends: python3.8-minimal (= 3.8.10-0ubuntu1~20.04.8)
  Depends: libpython3.8-stdlib (= 3.8.10-0ubuntu1~20.04.8)
  Depends: mime-support
python3.8-minimal
  PreDepends: libc6 (>= 2.29)
  Depends: libpython3.8-minimal (= 3.8.10-0ubuntu1~20.04.8)
  Depends: libexpat1 (>= 2.1~beta3)
  Depends: zlib1g (>= 1:1.2.0)
libpython3.8-stdlib
  Depends: libpython3.8-minimal (= 3.8.10-0ubuntu1~20.04.8)
  Depends: mime-support
  Depends: libbz2-1.0
  Depends: libc6 (>= 2.28)
  Depends: libcrypt1 (>= 1:4.1.0)
  Depends: libdb5.3
  Depends: libffi7 (>= 3.3~20180313)
  Depends: liblzma5 (>= 5.1.1alpha+20120614)
  Depends: libmpdec2
  Depends: libncursesw6 (>= 6)
  Depends: libreadline8 (>= 7.0~beta)
  Depends: libsqlite3-0 (>= 3.7.15)
  Depends: libtinfo6 (>= 6)
  Depends: libuuid1 (>= 2.20.1)
mime-support
libc6
  Depends: libgcc-s1
  Depends: libcrypt1 (>= 1:4.4.10-10ubuntu4)
libpython3.8-minimal
  Depends: libc6 (>= 2.4)
  Depends: libssl1.1 (>= 1.1.1)
libexpat1
  Depends: libc6 (>= 2.25)
zlib1g
  Depends: libc6 (>= 2.14)
libbz2-1.0
  Depends: libc6 (>= 2.4)
libcrypt1
  Depends: libc6 (>= 2.25)
libdb5.3
  Depends: libc6 (>= 2.17)
libffi7
  Depends: libc6 (>= 2.14)
liblzma5
  Depends: libc6 (>= 2.17)
libmpdec2
  Depends: libc6 (>= 2.14)
libncursesw6
  Depends: libtinfo6 (= 6.2-0ubuntu2.1)
  Depends: libc6 (>= 2.14)
libreadline8
  Depends: readline-common
  Depends: libc6 (>= 2.15)
  Depends: libtinfo6 (>= 6)
libsqlite3-0
  Depends: libc6 (>= 2.29)
libtinfo6
  Depends: libc6 (>= 2.16)
libuuid1
  Depends: libc6 (>= 2.25)
libgcc-s1
  Depends: libc6 (>= 2.14)
libssl1.1
  Depends: libc6 (>= 2.25)
```
  
</details>

### CODE REVIEW

**Comment by rebornplusplus** [comment]
cc @woky @cjdcordeiro @SamirPS

**Comment by rebornplusplus** [comment]
Small update: Added `readline-common` slices in https://github.com/canonical/chisel-releases/pull/44/commits/704ce026e29597e3ecb020530992d72345faf44f and rebased. Sorry for any inconvenience.

**Review by woky — APPROVED** [review]
Installation works. Thanks, LGTM.

**Review by SamirPS — APPROVED** [review]
LGTM

File: slices/libpython3.8-minimal.yaml
Comment (cjdcordeiro) [inline slices/libpython3.8-minimal.yaml:9]: if you leave this out, will it still work? I'm wondering if this should be in a separate config slice, cause it's not a library

File: slices/libpython3.8-stdlib.yaml
Comment (cjdcordeiro) [inline slices/libpython3.8-stdlib.yaml:1]: (same comment here as above)

Comment (cjdcordeiro) [reply slices/libpython3.8-stdlib.yaml:1]: a good example here would be `xml`...it could be its own slice, cause it is (?) self-contained and I could've save some space if I didn't have to install it

File: slices/python3.8-minimal.yaml
Comment (cjdcordeiro) [inline slices/python3.8-minimal.yaml:12]: if this from the maintainer scripts? If so, out of curiosity, could the same be done in starlark?

File: slices/python3.8-minimal.yaml
Comment (cjdcordeiro) [inline slices/python3.8-minimal.yaml:4]: I'm not sure this is the right name. we've previously used `core` for scenarios where you can extract a subset from a package that just "works" on its own. here, it is just a binary and everything else comes from the dependencies. To play the devil's advocate: _why would this be `core` while in openssl it is called `bins`?_

File: slices/readline-common.yaml
Comment (cjdcordeiro) [inline slices/readline-common.yaml:1]: what's this used for?

**Review by cjdcordeiro — CHANGES_REQUESTED** [review]
Great @rebornplusplus ! And thanks for the alphabetically ordered paths :) that helps with the review.

I left a few comments. At least one (minor) needs fixing

File: slices/libpython3.8-minimal.yaml
Comment (SamirPS) [reply slices/libpython3.8-minimal.yaml:1]: Doing this will maybe be difficult if one network lib depend on data lib no ?

File: slices/libbz2-1.0.yaml
Comment (rebornplusplus) [reply slices/libbz2-1.0.yaml:8]: Right, good catch. I actually forgot to omit the `x86_64` and `gnu` part.

File: slices/libpython3.8-minimal.yaml
Comment (rebornplusplus) [reply slices/libpython3.8-minimal.yaml:9]: So actually the `/usr/lib/python3.8/sitecustomize.py` library file is a symlink to this file. The `sitecustomize` module is imported in the `site` module, as per https://docs.python.org/3/library/site.html. 

However, any import error is silently ignored, thus this file can be removed if you really want to. But `/usr/lib/python3.8/sitecustomize.py` is a library file itself which is a symlink to the `/etc/python3.8/sitecustomize.py` file, that's why I kept that in the same slice.

Let me know if you still want me to relocate the `sitecustomize` file in a `config` slice and make that a dependency of this `libs` slice.

File: slices/python3.8-minimal.yaml
Comment (rebornplusplus) [reply slices/python3.8-minimal.yaml:4]: A'ight, sounds good. I will rename it to `bins`.

File: slices/readline-common.yaml
Comment (rebornplusplus) [reply slices/readline-common.yaml:1]: This package is a dependency of the `libreadline8` package, which is required for the [`readline` python module](https://docs.python.org/3.8/library/readline.html). 
The `readline-common` package provides only the configuration file `/etc/inputrc` which can and should be customized according to preference. See more here: https://tiswww.cwru.edu/php/chet/readline/rluserman.html#Readline-Init-File.
_(To make a vague comparison, you could say this is something like a default `.vimrc`)_

File: slices/libpython3.8-minimal.yaml
Comment (rebornplusplus) [reply slices/libpython3.8-minimal.yaml:1]: Logical grouping is indeed a good idea. But as Samir mentioned, it might be hard to do so with all the internal dependencies. However, I will give it a shot to check feasibility.

File: slices/libpython3.8-stdlib.yaml
Comment (rebornplusplus) [reply slices/libpython3.8-stdlib.yaml:1]: I am not sure if `xml` is self-contained. A quick grep inside the module revealed the following imports in the source. In any case, as I mentioned in the previous thread, I will try to granulate the slices. 

```log
root@124e3337e04c:~/libpython3.8-stdlib# grep -rnw "import" ./usr/lib/python3.8/xml/
./usr/lib/python3.8/xml/parsers/expat.py:2:import sys
./usr/lib/python3.8/xml/parsers/expat.py:4:from pyexpat import *
./usr/lib/python3.8/xml/sax/expatreader.py:8:from xml.sax._exceptions import *
./usr/lib/python3.8/xml/sax/expatreader.py:9:from xml.sax.handler import feature_validation, feature_namespaces
./usr/lib/python3.8/xml/sax/expatreader.py:10:from xml.sax.handler import feature_namespace_prefixes
./usr/lib/python3.8/xml/sax/expatreader.py:11:from xml.sax.handler import feature_external_ges, feature_external_pes
./usr/lib/python3.8/xml/sax/expatreader.py:12:from xml.sax.handler import feature_string_interning
./usr/lib/python3.8/xml/sax/expatreader.py:13:from xml.sax.handler import property_xml_string, property_interning_dict
./usr/lib/python3.8/xml/sax/expatreader.py:16:import sys
./usr/lib/python3.8/xml/sax/expatreader.py:22:    from xml.parsers import expat
./usr/lib/python3.8/xml/sax/expatreader.py:28:from xml.sax import xmlreader, saxutils, handler
./usr/lib/python3.8/xml/sax/expatreader.py:37:    import _weakref
./usr/lib/python3.8/xml/sax/expatreader.py:42:    import weakref
./usr/lib/python3.8/xml/sax/expatreader.py:442:    import xml.sax.saxutils
./usr/lib/python3.8/xml/sax/xmlreader.py:4:from . import handler
./usr/lib/python3.8/xml/sax/xmlreader.py:6:from ._exceptions import SAXNotSupportedException, SAXNotRecognizedException
./usr/lib/python3.8/xml/sax/xmlreader.py:116:        from . import saxutils
./usr/lib/python3.8/xml/sax/__init__.py:22:from .xmlreader import InputSource
./usr/lib/python3.8/xml/sax/__init__.py:23:from .handler import ContentHandler, ErrorHandler
./usr/lib/python3.8/xml/sax/__init__.py:24:from ._exceptions import SAXException, SAXNotRecognizedException, \
./usr/lib/python3.8/xml/sax/__init__.py:36:    import io
./usr/lib/python3.8/xml/sax/__init__.py:58:    import xml.sax.expatreader
./usr/lib/python3.8/xml/sax/__init__.py:60:import os, sys
./usr/lib/python3.8/xml/sax/__init__.py:82:            import sys
./usr/lib/python3.8/xml/sax/__init__.py:98:        from org.python.core import imp
./usr/lib/python3.8/xml/sax/saxutils.py:6:import os, urllib.parse, urllib.request
./usr/lib/python3.8/xml/sax/saxutils.py:7:import io
./usr/lib/python3.8/xml/sax/saxutils.py:8:import codecs
./usr/lib/python3.8/xml/sax/saxutils.py:9:from . import handler
./usr/lib/python3.8/xml/sax/saxutils.py:10:from . import xmlreader
./usr/lib/python3.8/xml/sax/saxutils.py:73:        import sys
./usr/lib/python3.8/xml/sax/_exceptions.py:2:import sys
./usr/lib/python3.8/xml/sax/_exceptions.py:4:    from java.lang import Exception
./usr/lib/python3.8/xml/etree/ElementInclude.py:51:import copy
./usr/lib/python3.8/xml/etree/ElementInclude.py:52:from . import ElementTree
./usr/lib/python3.8/xml/etree/ElementTree.py:95:import sys
./usr/lib/python3.8/xml/etree/ElementTree.py:96:import re
./usr/lib/python3.8/xml/etree/ElementTree.py:97:import warnings
./usr/lib/python3.8/xml/etree/ElementTree.py:98:import io
./usr/lib/python3.8/xml/etree/ElementTree.py:99:import collections
./usr/lib/python3.8/xml/etree/ElementTree.py:100:import collections.abc
./usr/lib/python3.8/xml/etree/ElementTree.py:101:import contextlib
./usr/lib/python3.8/xml/etree/ElementTree.py:103:from . import ElementPath
./usr/lib/python3.8/xml/etree/ElementTree.py:763:                    import locale
./usr/lib/python3.8/xml/etree/ElementTree.py:1500:            from xml.parsers import expat
./usr/lib/python3.8/xml/etree/ElementTree.py:1503:                import pyexpat as expat
./usr/lib/python3.8/xml/etree/ElementTree.py:1647:                from xml.parsers import expat
./usr/lib/python3.8/xml/etree/ElementTree.py:2058:    from _elementtree import *
./usr/lib/python3.8/xml/etree/ElementTree.py:2059:    from _elementtree import _set_factories
./usr/lib/python3.8/xml/etree/cElementTree.py:3:from xml.etree.ElementTree import *
./usr/lib/python3.8/xml/etree/ElementPath.py:55:# to import this module directly; the <b>ElementTree</b> does this for
./usr/lib/python3.8/xml/etree/ElementPath.py:59:import re
./usr/lib/python3.8/xml/dom/minidom.py:18:import io
./usr/lib/python3.8/xml/dom/minidom.py:19:import xml.dom
./usr/lib/python3.8/xml/dom/minidom.py:21:from xml.dom import EMPTY_NAMESPACE, EMPTY_PREFIX, XMLNS_NAMESPACE, domreg
./usr/lib/python3.8/xml/dom/minidom.py:22:from xml.dom.minicompat import *
./usr/lib/python3.8/xml/dom/minidom.py:23:from xml.dom.xmlbuilder import DOMImplementationLS, DocumentLS
./usr/lib/python3.8/xml/dom/minidom.py:1785:            raise xml.dom.NotSupportedErr("cannot import document nodes")
./usr/lib/python3.8/xml/dom/minidom.py:1787:            raise xml.dom.NotSupportedErr("cannot import document type nodes")
./usr/lib/python3.8/xml/dom/minidom.py:1958:        from xml.dom import expatbuilder
./usr/lib/python3.8/xml/dom/minidom.py:1961:        from xml.dom import pulldom
./usr/lib/python3.8/xml/dom/minidom.py:1968:        from xml.dom import expatbuilder
./usr/lib/python3.8/xml/dom/minidom.py:1971:        from xml.dom import pulldom
./usr/lib/python3.8/xml/dom/minicompat.py:7:# This module should only be imported using "import *".
./usr/lib/python3.8/xml/dom/minicompat.py:45:import xml.dom
./usr/lib/python3.8/xml/dom/__init__.py:140:from .domreg import getDOMImplementation, registerDOMImplementation
./usr/lib/python3.8/xml/dom/expatbuilder.py:30:from xml.dom import xmlbuilder, minidom, Node
./usr/lib/python3.8/xml/dom/expatbuilder.py:31:from xml.dom import EMPTY_NAMESPACE, EMPTY_PREFIX, XMLNS_NAMESPACE
./usr/lib/python3.8/xml/dom/expatbuilder.py:32:from xml.parsers import expat
./usr/lib/python3.8/xml/dom/expatbuilder.py:33:from xml.dom.minidom import _append_child, _set_attribute_node
./usr/lib/python3.8/xml/dom/expatbuilder.py:34:from xml.dom.NodeFilter import NodeFilter
./usr/lib/python3.8/xml/dom/domreg.py:9:import sys
./usr/lib/python3.8/xml/dom/domreg.py:45:    DOMImplementation object if the import succeeds.
./usr/lib/python3.8/xml/dom/domreg.py:52:    import os
./usr/lib/python3.8/xml/dom/pulldom.py:1:import xml.sax
./usr/lib/python3.8/xml/dom/pulldom.py:2:import xml.sax.handler
./usr/lib/python3.8/xml/dom/pulldom.py:18:        from xml.dom import XML_NAMESPACE
./usr/lib/python3.8/xml/dom/pulldom.py:161:            import xml.dom.minidom
./usr/lib/python3.8/xml/dom/pulldom.py:220:        import warnings
./usr/lib/python3.8/xml/dom/pulldom.py:343:    from io import StringIO
./usr/lib/python3.8/xml/dom/xmlbuilder.py:3:import copy
./usr/lib/python3.8/xml/dom/xmlbuilder.py:4:import warnings
./usr/lib/python3.8/xml/dom/xmlbuilder.py:5:import xml.dom
./usr/lib/python3.8/xml/dom/xmlbuilder.py:7:from xml.dom.NodeFilter import NodeFilter
./usr/lib/python3.8/xml/dom/xmlbuilder.py:194:            import urllib.request
./usr/lib/python3.8/xml/dom/xmlbuilder.py:204:        import xml.dom.expatbuilder
./usr/lib/python3.8/xml/dom/xmlbuilder.py:227:        import posixpath, urllib.parse
./usr/lib/python3.8/xml/dom/xmlbuilder.py:246:        import urllib.request
```

File: slices/libpython3.8-minimal.yaml
Comment (cjdcordeiro) [reply slices/libpython3.8-minimal.yaml:9]: ah ok. if it's a symlink to a lib then nevermind ;) ty

File: slices/libpython3.8-minimal.yaml
Comment (cjdcordeiro) [reply slices/libpython3.8-minimal.yaml:1]: y it only makes sense if the slice can be self-contained. i.e. it is reasonable for each slice to rely on a `minimal` package and/or a `core` slice of its own package, but if it can be self-contained, then it makes sense to slice it further. maybe we should do this exercise with the `stdlib` package first as that's a more logical candidate for self-contained slices, and see what we learn from that one.

File: slices/libpython3.8-stdlib.yaml
Comment (cjdcordeiro) [reply slices/libpython3.8-stdlib.yaml:1]: at first glance, **I think** this slice would just depend on `libpython3.8-minimal_libs`. This is an interesting exercise. Lemme know if you need help.

P.S.: this is a good reference example: https://github.com/canonical/chisel-releases/blob/ubuntu-22.04/slices/openjdk-8-jre-headless.yaml

Comment (cjdcordeiro) [reply slices/libpython3.8-stdlib.yaml:1]: (unresolving so we can track progress on this thread)

File: slices/readline-common.yaml
Comment (cjdcordeiro) [reply slices/readline-common.yaml:1]: > If that file does not exist or cannot be read, the ultimate default is /etc/inputrc.

what if `/etc/inputrc` doesn't exist :hand_over_mouth: ?

**Comment by rebornplusplus** [comment]
Hiya! I have added two new commits (https://github.com/canonical/chisel-releases/pull/44/commits/6b99feaefcd39a8325ff7fc0cb5c20a104bf21b7, https://github.com/canonical/chisel-releases/pull/44/commits/8867328733b01cff8dd8696124a24788b94fcbc3) which spans out the `libs` slice of `libpython3.8-stdlib` and splits into multiple logical groups.
I have taken inspiration from the [Python 3.8 Standard Library documentation](https://docs.python.org/3.8/library/index.html) to do the grouping.

However, I couldn't manage to do the same for the `libpython3.8-minimal` package. Most of the libraries in there are very tightly coupled. I started with the grouping but later found out the dependencies among those slices were insane, resulting in a big dependency circle. It remains as it was.

I have tested all of the slices as best I could, but I would not be too confident. They should work. :crossed_fingers:

File: slices/libpython3.8-stdlib.yaml
Comment (cjdcordeiro) [inline slices/libpython3.8-stdlib.yaml:265]:
```suggestion
  markup-tools:
```

(to be consistent with "development tools")

File: slices/libpython3.8-stdlib.yaml
Comment (cjdcordeiro) [inline slices/libpython3.8-stdlib.yaml:324]:
```suggestion
  data-persistence:
```

(being explicit helps here)

File: slices/libpython3.8-stdlib.yaml
Comment (cjdcordeiro) [inline slices/libpython3.8-stdlib.yaml:374]: can you add a quick comment with the rationale for having this extra slice composed of outliers and why they are outliers?

File: slices/python3.8.yaml
Comment (cjdcordeiro) [inline slices/python3.8.yaml:1]: do we need another slice here? The thing is that we should have a way to say `chisel cut ... python3.8_<slice>` such that we get python3.8 without having to repeat all the slice names from `libpython3.8-stdlib`. Otherwise, this package becomes irrelevant as it only provides the core python slices, and we'll end up having to list all the slices from `libpython3.8-stdlib`. Maybe a `stdlib` slice that depends on `python3.8_core` + all stdlib slices?

File: slices/libpython3.8-stdlib.yaml
Comment (cjdcordeiro) [reply slices/libpython3.8-stdlib.yaml:1]: current slicing looking good

File: slices/libpython3.8-stdlib.yaml
Comment (rebornplusplus) [reply slices/libpython3.8-stdlib.yaml:217]:
```
$ find libpython3.8-stdlib -type f | xargs ldd | grep uuid
libpython3.8-stdlib/usr/lib/python3.8/lib-dynload/_uuid.cpython-38-x86_64-linux-gnu.so:
	libuuid.so.1 => /lib/x86_64-linux-gnu/libuuid.so.1 (0x00007f1d773f0000)
```

```
$ grep -rni "libuuid" libpython3.8-stdlib
Binary file libpython3.8-stdlib/usr/lib/python3.8/lib-dynload/_uuid.cpython-38-x86_64-linux-gnu.so matches
libpython3.8-stdlib/usr/lib/python3.8/uuid.py:602:        # The uuid_generate_* routines are provided by libuuid on at least
```

Additionally, I found no module using `dlopen()` except `ctypes`, but found no mention of `libuuid` or `uuid` there.

File: slices/libpython3.8-stdlib.yaml
Comment (rebornplusplus) [reply slices/libpython3.8-stdlib.yaml:1]: Yes. I will soon add a `python3.8_standard` slice to include all of them, per your suggestion.

File: slices/libpython3.8-stdlib.yaml
Comment (rebornplusplus) [reply slices/libpython3.8-stdlib.yaml:316]: The `/usr/lib/python3.8/_osx_support.py` file didn't fit anywhere else, and it is important enough to not be in `extras` since some slices do depend on it. That's why I created one for this one. Let me know if you want to rename it to something else.

File: slices/libpython3.8-minimal.yaml
Comment (rebornplusplus) [reply slices/libpython3.8-minimal.yaml:1]: Added in https://github.com/canonical/chisel-releases/pull/44/commits/a5a399095ebe12134a82261a91529998a932c703.

File: slices/libpython3.8-minimal.yaml
Comment (rebornplusplus) [reply slices/libpython3.8-minimal.yaml]: Yup, you are right. Moved it in https://github.com/canonical/chisel-releases/pull/44/commits/a5a399095ebe12134a82261a91529998a932c703.

File: slices/libpython3.8-stdlib.yaml
Comment (rebornplusplus) [reply slices/libpython3.8-stdlib.yaml:376]: Added in https://github.com/canonical/chisel-releases/pull/44/commits/59ea763199ebca5f334753d40b701d828b7ea99a.

File: slices/python3.8.yaml
Comment (rebornplusplus) [reply slices/python3.8.yaml:1]: Added `python3.8_standard` in https://github.com/canonical/chisel-releases/pull/44/commits/bb3524108df560653f1863091df70dbc1ab2defd.

**Comment by rebornplusplus** [comment]
Added new commits. Let me know what you think. If possible, I would like to rebase the commits before this goes in, so that each file is affected by exactly one commit.

**Review by cjdcordeiro — APPROVED** [review]
:partying_face: greeeaatt

thanks @rebornplusplus 

LGTM. Feel free to rebase it and let me know once I can merge it.

**Comment by cjdcordeiro** [comment]
@woky can you pls re-review? (there were several changes after your last review)

**Review by woky — APPROVED** [review]
I'm not sure about the slices in `-stdlib`. E.g.
```
  filesys:
    essential:
      - libpython3.8-stdlib_core
    contents:
      /usr/lib/python3.8/filecmp.py:
      /usr/lib/python3.8/fileinput.py:
```
What makes `fileinput.py` more _filesys_ and less _cli_?

```
  # Program Frameworks
  # https://docs.python.org/3.8/library/frameworks.html
  frameworks:
    essential:
      - libpython3.8-stdlib_core
      - libpython3.8-stdlib_text
    contents:
      /usr/lib/python3.8/cmd.py:
      /usr/lib/python3.8/shlex.py:
      /usr/lib/python3.8/turtle.py:
```
What makes `shlex.py` a _framework_ compared to other modules?

I'm not going to bike-shed over categories. Ideally, we would slice individual modules, but that'd take more effort. Since I'm late in the review and this slicing will allow people to get what they want, I don't see any other problems. LGTM.

**Comment by rebornplusplus** [comment]
> What makes `fileinput.py` more _filesys_ and less _cli_?
> 
> What makes `shlex.py` a _framework_ compared to other modules?

Yeah I agree, the categories are a bit shaky at the moment. Individual modules would probably have been the better way to go, but doesn't seem very reasonable to write down a slice per module. I took the categorization from https://docs.python.org/3.8/library/index.html and would like to point the finger at those docs. :stuck_out_tongue:

**Comment by woky** [comment]
> > What makes `fileinput.py` more _filesys_ and less _cli_?
> > What makes `shlex.py` a _framework_ compared to other modules?
> 
> Yeah I agree, the categories are a bit shaky at the moment. Individual modules would probably have been the better way to go, but doesn't seem very reasonable to write down a slice per module. I took the categorization from https://docs.python.org/3.8/library/index.html and would like to point the finger at those docs. 😛

Oh, OK. That makes it clear. :-) Thanks.

**Pull request merged**
